src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -122,22 +131,66 @@ internal void UpdateHtmlBuffer(JToken token)
                 return;
             }
 
-            var hostDocumentUri = ConvertFilePathToUri(request.HostDocumentFilePath);
+            var hostDocumentUri = new Uri(request.HostDocumentFilePath);
             _documentManager.UpdateVirtualDocument<HtmlVirtualDocument>(
                 hostDocumentUri,
                 request.Changes,
                 request.HostDocumentVersion);
         }
 
-        private static Uri ConvertFilePathToUri(string filePath)
+        public override async Task<RazorDocumentRangeFormattingResponse> RazorRangeFormattingAsync(RazorDocumentRangeFormattingParams request, CancellationToken cancellationToken)
         {
-            if (filePath.StartsWith("/", StringComparison.Ordinal) && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            var response = new RazorDocumentRangeFormattingResponse() { Edits = Array.Empty<TextEdit>() };
+
+            if (request.Kind == RazorLanguageKind.Razor)
+            {
+                return response;
+            }
+
+            await _joinableTaskFactory.SwitchToMainThreadAsync();
+            
+            var hostDocumentUri = new Uri(request.HostDocumentFilePath);
+            if (!_documentManager.TryGetDocument(hostDocumentUri, out var documentSnapshot))
+            {
+                return response;
+            }
+
+            var serverKind = default(LanguageServerKind);
+            var projectedUri = default(Uri);
+            if (request.Kind == RazorLanguageKind.CSharp &&
+                documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDocument))
+            {
+                serverKind = LanguageServerKind.CSharp;
+                projectedUri = csharpDocument.Uri;
+            }
+            else if (request.Kind == RazorLanguageKind.Html &&
+                documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlDocument))
+            {
+                serverKind = LanguageServerKind.Html;
+                projectedUri = htmlDocument.Uri;
+            }
+            else
             {
-                filePath = filePath.Substring(1);
+                Debug.Fail("Unexpected RazorLanguageKind. This can't really happen in a real scenario.");
+                return response;
             }
 
-            var uri = new Uri(filePath, UriKind.Absolute);
-            return uri;
+            var formattingParams = new DocumentRangeFormattingParams()
+            {
+                TextDocument = new TextDocumentIdentifier() { Uri = projectedUri },
+                Range = request.ProjectedRange,
+                Options = request.Options
+            };
+
+            var edits = await _requestInvoker.RequestServerAsync<DocumentRangeFormattingParams, TextEdit[]>(
```

Comment:
Is there any harm in assigning this directly to `response.Edits` on the next line?

Reply:
Nope. This is what happens when I refactor the code multiple times 😄